proposal adder helper

adibaejaz opened the pull request:
Pull request for projectum 20ae_newproposaladder
closes #361

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -34,45 +34,51 @@ def subparser(subpi):
                        help="Url for call for proposals"
                        )
-    subpi.add_argument("-cpp", "--cpp_info", nargs="+",
+    subpi.add_argument("--cpp_info", nargs="+",
```
this will be too hard for me because I don't know the keys off the top of my head.

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -34,45 +34,51 @@ def subparser(subpi):
                        help="The end date for the proposed grant in format YYYY-MM-DD."
                        )
-    subpi.add_argument("--call", "--call_for_proposals", nargs="+",
+    subpi.add_argument("--call_for_proposals", nargs="+",
```
we should probably talk about which of all the many variables are needed in the CLI.....not all for sure.

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -119,10 +124,10 @@ def construct_global_ctx(self):
     def db_updater(self):
         rc = self.rc
-        if not rc.date:
+        if not rc.submitted_date:
```
please don't auto-set this

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -132,45 +137,63 @@ def db_updater(self):
-            pdoc.update({'group_members': rc.group_members})
+            pdoc.update({'duration': 'tbd'})
+        pdoc.update({
+            'end_day': rc.end_day,
```
let's figure out how to handle dates....

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -132,45 +137,63 @@ def db_updater(self):
+            pdoc.update({'pi': rc.pi})
+        else:
+            pdoc.update({'pi': 'tbd'})
+        if rc.status:
```
auto-set this to ``inprep``
remove from CLI

adibaejaz replied:
are you referring to `pi` or `status`?

sbillinge replied:
comments always appear below the line they refer to....

adibaejaz commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -34,45 +34,51 @@ def subparser(subpi):
                        )
-    subpi.add_argument("--begin_date", nargs="+",
-                       help="The start date for the proposed grant in format YYYY-MM-DD."
+    subpi.add_argument("--begin_day", nargs="+",
```
does this mean we only allow the user to input a complete begin date, and not, for example, only a begin year?

sbillinge replied:
yup.  For proposals this is needed.

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -28,70 +23,57 @@ def subparser(subpi):
                        )
-    subpi.add_argument("-a", "--authors",
-                       help="Other investigator names"
+    subpi.add_argument("authors", help="Other investigator names"
```
this will need an ``nargs='+'`` so a list can be input

adibaejaz replied:
done

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -28,70 +23,57 @@ def subparser(subpi):
                        )
-    subpi.add_argument("--end_date", nargs="+",
-                       help="The end date for the proposed grant in format YYYY-MM-DD."
+    subpi.add_argument("submitted_date", help="Day on which the proposal is submitted "
```
please make optional. In general I won't know this when I create the proposal so will rarely, if ever, use it.  In fact, maybe delete it from the arg list?

adibaejaz replied:
done

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -28,70 +23,57 @@ def subparser(subpi):
                        )
-    subpi.add_argument("--call_for_proposals", nargs="+",
-                       help="Url for call for proposals"
+    subpi.add_argument("pi", help="Name of principal investigator"
```
please make optional.  by default have it find the group pi name from the rc.  Please see how this is done in the projectum adder

adibaejaz replied:
Will do! Just to clarify, the projectum adder displays the `pi_id` whereas here I am trying to display the `pi` name. Should I change my adder to display `pi_id`? 
Alternately, I have looked the `get_pi_id(rc)` function in `regolith/tools.py` and I can write another function `get_pi_name(rc)` that returns the `pi` name, if that's a better option. Unless, do we have a function in regolith that returns the name of a group member given their id?

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -28,70 +23,57 @@ def subparser(subpi):
-    subpi.add_argument("-c", "--currency",
-                       help="Currency in which amount is specified."
-                            " Typically $ or USD"
+    subpi.add_argument("-b", "--begin_date", nargs="+",
```
I think make this required, and don't narg it, there will only be one begin date, by definition

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -28,70 +23,57 @@ def subparser(subpi):
                        )
-    subpi.add_argument("--due_date", nargs="+",
-                       help="The due date for the proposal in format YYYY-MM-DD."
+    subpi.add_argument("--cpp_agencies", nargs="+",
```
no need to hvae the schema structure leak into the arg, so I would just call this ``other_agencies``

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -28,70 +23,57 @@ def subparser(subpi):
                        )
-    subpi.add_argument("-d", "--duration",
-                       help="Duration of proposal in years"
+    subpi.add_argument("--cpp_institution", nargs="+",
```
no cpp

adibaejaz replied:
done

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -28,70 +23,57 @@ def subparser(subpi):
                        )
-    subpi.add_argument("--funder", nargs="+",
-                       help="Who funds the proposal. As funder in grants"
+    subpi.add_argument("--cpp_months_academic", nargs="+",
```
no nargs, this is not a list.  no cpp

adibaejaz replied:
done

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -28,70 +23,57 @@ def subparser(subpi):
-    subpi.add_argument("--funder", nargs="+",
-                       help="Who funds the proposal. As funder in grants"
+    subpi.add_argument("--cpp_months_academic", nargs="+",
+                       help="Number of working months in the academic year"
```
this is very good help, thanks!

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -28,70 +23,57 @@ def subparser(subpi):
                        )
-    subpi.add_argument("--full", nargs="+",
-                       help="Full body of the proposal"
+    subpi.add_argument("--cpp_months_summer", nargs="+",
```
see above, same issue

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -28,70 +23,57 @@ def subparser(subpi):
                        )
-    subpi.add_argument("-n", "--notes", nargs="+",
-                       help="Anything to note"
+    subpi.add_argument("--cpp_scope", nargs="+",
```
no cpp, no nargs

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -28,70 +23,57 @@ def subparser(subpi):
-    subpi.add_argument("-m", "--team_members", nargs="+",
-                       help="Information about the team members participating"
-                            "in the grant"
+    subpi.add_argument("-e","--end_date", nargs="+",
```
no nargs, make not optional

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -28,70 +23,57 @@ def subparser(subpi):
                        )
-    subpi.add_argument("-t", "--title",
-                        help="Actual title of the proposal"
+    subpi.add_argument("-f", "--funder", nargs="+",
```
no nargs

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -28,70 +23,57 @@ def subparser(subpi):
-    subpi.add_argument("-t", "--title",
-                        help="Actual title of the proposal"
+    subpi.add_argument("-f", "--funder", nargs="+",
+                       help="Who funds the proposal. As funder in grants"
```
"agency where the proposal is being submitted"

adibaejaz replied:
done

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -124,11 +106,8 @@ def construct_global_ctx(self):
-        else:
-            now = date_parser.parse(rc.submitted_date).date()
-        key = f"{str(now.year)[2:]}{rc.lead[:2]}_{''.join(rc.name.casefold().split()).strip()}"
+        submitted_date = date_parser.parse(rc.submitted_date).date()
```
please use the old logic unchanged, so it is now-date that is used for the key

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -148,73 +127,65 @@ def db_updater(self):
-            'call_for_proposals': rc.call_for_proposals,
-            # 'cpp_info':
-        })
+            pdoc.update({'authors': ['tbd']})
```
empty string please, your version is introducing "magic"

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -148,73 +127,65 @@ def db_updater(self):
-            pdoc.update({'status': 'tbd'})
-        # pdoc.update({'team':})
+        pdoc.update({'status': 'inprep'})
+        sample_team = {'cv':'http://pdf.com/goodbye-cv',
```
no magic please, just empty list for team if not specified

adibaejaz replied:
I used this as a "sample team member", sort of like the sample milestone in the projectum adder, which does not allow the user to pass in a milestone but instead provides a template that can be edited and replicated. I will change it to an empty list. Do you think I should allow the user to pass in a list of team members, each with their {cv, name, email, institution, name, position, subaward_amount} ? The way I can think of going about this is either
a) define arguments `member1_cv,` `member2_cv,...` up to `member10_cv` and so on for `name,` `email,` `institution...`
b) allow the user to pass in a list of `cv`'s, a list of `name`s, etc. for the team members and then construct a member dictionary by iterating through these lists, i.e. `cv[0],` `names[0],` etc. would be assigned to the first member, `cv[1],` `names[1],` etc. to the second and so on. But this method can be problematic when the lists for each key type are not aligned, for e.g. member 1 specifies a `cv[0]` but member 2 does not, and `cv[1]` is actually the `cv` of member 3. So we would instruct the user to specify `None` for no cv 
I'm starting to think it might be easier for the user to simply edit the `proposals.yaml` file themselves for the `team` key!

sbillinge replied:
thanks.  Agile says no for now (but be careful, I may be requesting it when I decide I really need it!)

sbillinge commented on tests/test_helpers.py:
```diff
@@ -17,6 +17,10 @@
       "--date", "2020-04-29", "-c", "afriend", "-d", "more work",
       "-m", "ascopatz", "-g", "SymPy-1.1"],
      "20ly_newprojectum has been added in projecta\n"),
+    (["helper", "a_proposal", "a new proposal", 500.0, "Kurt Godel", "USD"
```
maybe make 2 or 3 tests that test reasonable edge cases (but don't go crazy to test every tiny possibility in this case)

adibaejaz commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -28,70 +23,57 @@ def subparser(subpi):
                        )
-    subpi.add_argument("--submitted_date",
-                       help="Day on which the proposal is submitted in format YYYY-MM-DD"
+    subpi.add_argument("--cpp_flag", nargs="+",
```
sorry about all the unnecessary nargs functionality, this was an oversight while doing some copy-pasting

sbillinge commented:
we always hand around id's.  If you want to display a name you dereference
it using ``fuzzy_retrieval()`` before formulating the print statement

On Wed, Jun 3, 2020 at 11:06 AM Adiba Ejaz <notifications@github.com> wrote:

> *@adibaejaz* commented on this pull request.
> ------------------------------
>
> In regolith/helpers/a_proposalhelper.py
> <https://github.com/regro/regolith/pull/372#discussion_r434638497>:
>
> >                         )
> -    subpi.add_argument("--call_for_proposals", nargs="+",
> -                       help="Url for call for proposals"
> +    subpi.add_argument("pi", help="Name of principal investigator"
>
> Will do! Just to clarify, the projectum adder displays the pi_id whereas
> here I am trying to display the pi name. Should I change my adder to
> display pi_id?
> Alternately, I have looked the get_pi_id(rc) function in regolith/tools.py
> and I can write another function get_pi_name(rc) that returns the pi
> name, if that's a better option. Unless, do we have a function in regolith
> that returns the name of a group member given their id?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/regro/regolith/pull/372#discussion_r434638497>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAOWUJGRM7YIUWYK7PB3HDRUZRFHANCNFSM4NOLRPHA>
> .
>


-- 
Professor Simon Billinge
Columbia University

sbillinge commented:
thanks, great questions...

On Wed, Jun 3, 2020 at 11:23 AM Adiba Ejaz <notifications@github.com> wrote:

> I used this as a "sample team member", sort of like the sample milestone
> in the projectum adder, which does not allow the user to pass in a
> milestone but instead provides a template that can be edited and
> replicated. I will change it to an empty list. Do you think I should allow
> the user to pass in a list of team members, each with their {cv, name,
> email, institution, name, position, subaward_amount} ? The way I can think
> of going about this is either
>
yes, understood.  We just want to avoid "wrong" data in the database, so
unpopulated (but not blank) fields are better if  we lack info.  In answer
to your question, I think no.  This is better done in the editor probably.
Agile says don't implement functionality, especially complicated
functionality, until someone asks for it......

> a) define arguments member1_cv, member2_cv,... up to member10_cv and so
> on for name, email, institution...
> b) allow the user to pass in a list of cv's, a list of names, etc. for
> the team members and then construct a member dictionary by iterating
> through these lists, i.e. cv[0], names[0], etc. would be assigned to the
> first member, cv[1], names[1], etc. to the second and so on. But this
> method can be problematic when the lists for each key type are not aligned,
> for e.g. member 1 specifies a cv[0] but member 2 does not, and cv[1] is
> actually the cv of member 3. So we would instruct the user to specify None
> for no cv
> I'm starting to think it might be easier for the user to simply edit the
> proposals.yaml file themselves for the team key!
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/regro/regolith/pull/372#discussion_r434651491>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAOWUK5CXVZU432GTGLSSLRUZTHPANCNFSM4NOLRPHA>
> .
>


-- 
Professor Simon Billinge
Columbia University

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -0,0 +1,197 @@
+    subpi.add_argument("name", help="A short but unique name for the proposal",
+                       default=None
+                       )
+    subpi.add_argument("amount", help="value of award",
```
these are required args, so I don't think they should take defaults, at least I can't see how they would ever take the default value since the code won't run if they are not specified.

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -0,0 +1,197 @@
+    subpi.add_argument("amount", help="value of award",
+                       default=None
+                       )
+    subpi.add_argument("authors", nargs="+",
```
This should be optional I htink

adibaejaz replied:
done

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -0,0 +1,197 @@
+    subpi.add_argument("end_date", help="The end date for the proposed grant "
+                                        "in format YYYY-MM-DD."
+                       )
+    subpi.add_argument("due_date", help="The due date for the proposal in format YYYY-MM-DD."
```
can we put this one earlier, right after ``name``?

adibaejaz replied:
done

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -0,0 +1,197 @@
+    subpi.add_argument("authors", nargs="+",
+                       help="Other investigator names"
+                       )
+    subpi.add_argument("date", help="The begin date for the proposed grant "
```
please call it begin_date

adibaejaz replied:
done

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -0,0 +1,197 @@
+                       )
+    subpi.add_argument("due_date", help="The due date for the proposal in format YYYY-MM-DD."
+                       )
+    subpi.add_argument("duration", help="Duration of proposal in years"
```
first, I think that this should be in months.  Are you sure it is in years?   Second, if we know the begin and end-date this can be computed. Maybe request begin_date and right after, request duration, then compute the end-date internally.

adibaejaz replied:
done

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -0,0 +1,197 @@
+                       )
+    subpi.add_argument("-p", "--pi",
+                       help="Name of principal investigator. Defaults to"
+                            "group pi name in database"
```
I think "defaults to group pi name in regolithrc.json

adibaejaz replied:
done

sbillinge replied:
I don't see that

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -0,0 +1,197 @@
+                       help="Name of principal investigator. Defaults to"
+                            "group pi name in database"
+                       )
+    subpi.add_argument("--cppflag", help="Current and pending form (true or false)"
```
default to true

adibaejaz replied:
done

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -0,0 +1,197 @@
+    subpi.add_argument("--other_agencies", help="Other agencies to which the proposal has been "
+                                                "submitted"
+                       )
+    subpi.add_argument("--institution", help="The institution where the work will primarily"
```
There is a tweak to the schema I think in an unmerged PR (not your fault) and this is now a list.  Default to empty list

adibaejaz replied:
donedone

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -0,0 +1,197 @@
+    subpi.add_argument("--institution", help="The institution where the work will primarily"
+                                             "be carried out"
+                       )
+    subpi.add_argument("--months_academic", help="Number of working months in the academic year"
```
default to 0

adibaejaz replied:
done

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -0,0 +1,197 @@
+    subpi.add_argument("--months_academic", help="Number of working months in the academic year"
+                                                 "to be stated on the current and pending form"
+                       )
+    subpi.add_argument("--months_summer", help="Number of working months in the summer to be"
```
default to 0

adibaejaz replied:
done

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -0,0 +1,197 @@
+    subpi.add_argument("--months_summer", help="Number of working months in the summer to be"
+                                                "stated on the current and pending form"
+                       )
+    subpi.add_argument("--scope", help="Scope of the project"
```
default to empty string. help to "Scope of project and statement of any overlaps with other current and pending grants"

adibaejaz replied:
done

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -0,0 +1,197 @@
+                       )
+    subpi.add_argument("--scope", help="Scope of the project"
+                       )
+    subpi.add_argument("-f", "--funder",
```
default to empty string

adibaejaz replied:
done

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -0,0 +1,197 @@
+    def db_updater(self):
+        gtx = self.gtx
+        rc = self.rc
+        if not rc.date:
```
we are not asking for date in the CLI so just use today's date.

adibaejaz replied:
will do – would you prefer that the suffix to the key, for example, `20_proposal`, be the year from the `begin_date` or `now.year` ?

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -0,0 +1,197 @@
+        if rc.amount:
+            pdoc.update({'amount': rc.amount})
+        else:
+            pdoc.update({'amount': ''})
```
should be a number, not a string.  It may be better to have argparser to set the default values, I am not sure.  I know I have done it using this method too.  Maybe test it both ways, but it will make the code cleaner if the defaults from argparse works.

adibaejaz replied:
Is this conditional in my code is redundant because `amount` is a required argument hence the code won't run if it is not specified? If this is the case, I will remove the `if` condition on all required arguments. 
Are you suggesting that I make `amount` an optional argument and set  `default = 0` for it in the `argparse`? In that case, if my understanding is correct:
1.  if the user specifies `--amount 50` the output `amount: 50` will be printed
2. if the user specifies only `--amount` the output `amount: 0` will be printed
3. if the user does not specify --amount`, no `amount` key-value pair will be printed

sbillinge replied:
yes, the first thing.

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -0,0 +1,197 @@
+            else:
+                pdoc.update({'authors': rc.authors})
+        else:
+            pdoc.update({'authors': ['']})
```
just empty list is good, ``[]``

adibaejaz replied:
done

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -0,0 +1,197 @@
+        if rc.cppflag:
+            cpp_info['cppflag'] = rc.cppflag
+        if rc.other_agencies:
+            cpp_info['other_agencies_submitted'] = rc.other_agencies
```
These all need default values, but as discussed above, experiment with not changing code here but using defaults above.  Same comments apply below.

adibaejaz replied:
done

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -0,0 +1,197 @@
+        if rc.notes:
+            pdoc.update({'notes': rc.notes})
+        if rc.pi:
+            pdoc.update({'pi': rc.pi})
```
you seem to be handling this differently whether rc.pi exists or not, i.e., if pi is passed in you don't dereference it but if it is not then you do.  In general I think the best is to store ids in the database and only dereference them when we build something.

adibaejaz replied:
done – I handled them differently because the`pi` help asks for the pi name, not the pi_id, so there is no need to retrieve if the user has specified a `pi`
but I will now simply change the code to display the `pi_id` in both cases, and change the `pi` help to "ID of principal investigator"

sbillinge replied:
perfect, that is the behavior I want.

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -0,0 +1,197 @@
+            pi_name = pi_entry['name']
+            pdoc.update({'pi': pi_name})
+        pdoc.update({'status': 'inprep'})
+        sample_team = {'cv':'',
```
You took in authors.  It is a philosophical question if we need an author list and a team list or if they do the same job.  maybe we should just not use the authors list, and make ``team`` to look like 
```
team:
  name:
  subaward_amount:
```
and default ``subaward_amount`` to 0.    
This format is a bit historical.  Now we have contacts.yml we can dereference all the info if you just store the name, so we should never be putting email, institution, position and so on here.

adibaejaz replied:
done

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -0,0 +1,197 @@
+                            "first database in the regolithrc.json file."
+                       )
+    subpi.add_argument("name", help="A short but unique name for the proposal",
+                       default=None
```
these still seem to have default values

adibaejaz commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -28,70 +23,57 @@ def subparser(subpi):
-    subpi.add_argument("-d", "--duration",
-                       help="Duration of proposal in years"
+    subpi.add_argument("--cpp_institution", nargs="+",
+                       help="The institution to be stated on the current"
```
done

sbillinge commented on tests/test_helpers.py:
```diff
@@ -17,9 +17,9 @@
-    (["helper", "a_proposal", "a new proposal", "2020-12-29",
-      "500.0","2020-09-22","3", "To destroy numbers",
-      "-a", "Kurt Godel", "-p", "Simon Billinge", "-f", "Columbia"],
+    (["helper", "a_proposal", "a new proposal","500.0","To destroy numbers",
```
this helper is not very well tested.  At least I think we need two tests, a minimal set of inputs (to test the defaults are working correctly), and a maximal set (so we have coverage for everything).

sbillinge commented on tests/test_helpers.py:
```diff
@@ -17,6 +17,19 @@
      "20ly_newprojectum has been added in projecta\n"),
+    (["helper", "a_proposal", "an empty new proposal","100.0","To understand numbers"],
+    "20_an_empty_new_proposal has been added in proposals\n"),
+    (["helper", "a_proposal", "a sparse new proposal","500.0","To destroy numbers",
```
something I don't expect here is that an optional argument has been stated but no value given. This is not quite the behhavior I want, so if this is what our use of defaults is giving maybe we need to walk it back again.  I kind of want that if a flag is not specified it gets a default, but if it is specified it should take a value.      I may have just caused you lots of unnecessary work and we have to go back to the if ``rc.something`` wrapping....sorry

adibaejaz replied:
will do! luckily, my commits can save me here

sbillinge replied:
👍

sbillinge commented on tests/test_helpers.py:
```diff
@@ -17,6 +17,19 @@
+    (["helper", "a_proposal", "a dense new proposal","1000.0","To create numbers",
+      "--begin_date", "2020-09-22", "--duration", "3", "--due_date", "2020-12-29",
+      "-a", "Kurt Godel", "-c", "Bitcoin" "-p", "sbillinge", "--cppflag", "False",
+      "--other_agencies", "Flatland", "-i", "Columbia", "--months_academic", "6",
```
since the tests act a bit like examples, could you give the institution as columbiau (institution-id) which is what I want to put in there the most.  Also, is end_date missing?

adibaejaz replied:
will do for the institution. the `end_date` is internally computed using 

`begin_date = date_parser.parse(rc.begin_date).date()
  pdoc.update({'end_date': begin_date + relativedelta(months = rc.duration)})`

sbillinge replied:
yes you are so right....my apologies!

adibaejaz replied:
no problem!

sbillinge commented:
ah yes, you are right!  I am going mad.  THanks so much.  Get started also
on the new prum, as some of these last things to get this merged are slow
and tedious but not much coding, so having them in the background of
something else is better.

On Fri, Jun 5, 2020 at 3:58 PM Adiba Ejaz <notifications@github.com> wrote:

> *@adibaejaz* commented on this pull request.
> ------------------------------
>
> In tests/test_helpers.py
> <https://github.com/regro/regolith/pull/372#discussion_r436136655>:
>
> > @@ -17,6 +17,19 @@
>        "--date", "2020-04-29", "-c", "afriend", "-d", "more work",
>        "-m", "ascopatz", "-g", "SymPy-1.1"],
>       "20ly_newprojectum has been added in projecta\n"),
> +    (["helper", "a_proposal", "an empty new proposal","100.0","To understand numbers"],
> +    "20_an_empty_new_proposal has been added in proposals\n"),
> +    (["helper", "a_proposal", "a sparse new proposal","500.0","To destroy numbers",
> +      "-a", "-c", "-p", "--cppflag", "--other_agencies", "-i",
> +      "--months_academic", "--months_summer", "-s", "-f", "-n", "this is a minimal proposal"],
> +     "20_a_sparse_new_proposal has been added in proposals\n"),
> +    (["helper", "a_proposal", "a dense new proposal","1000.0","To create numbers",
> +      "--begin_date", "2020-09-22", "--duration", "3", "--due_date", "2020-12-29",
> +      "-a", "Kurt Godel", "-c", "Bitcoin" "-p", "sbillinge", "--cppflag", "False",
> +      "--other_agencies", "Flatland", "-i", "Columbia", "--months_academic", "6",
>
> will do for the institution. the end_date is internally computed using
>
> begin_date = date_parser.parse(rc.begin_date).date()
> pdoc.update({'end_date': begin_date + relativedelta(months = rc.duration)})
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/regro/regolith/pull/372#discussion_r436136655>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAOWUOZVE224EMTPQTZD5DRVFE5DANCNFSM4NOLRPHA>
> .
>


-- 
Professor Simon Billinge
Columbia University

adibaejaz commented:
I've pushed a functioning version of the adder with a cleaner use of defaults instead of if... else blocks! As you said, I've added both a minimal and maximal test proposal; the tests each pass individually provided we delete the other output from proposals.yml. I will work on the news now! Is the docstring alright? And should I be concerned that the Travis CI build has failed here? I will also get started on the new prum.

adibaejaz commented:
@sbillinge I have added the news – ready for final review!

sbillinge commented:
@adibaejaz I am afraid this seems to be failing tests.  Let me know if you need help to fix it.

adibaejaz commented:
@sbillinge If I'm not wrong, it's failing the tests because two tests for 2 different proposals have been specified, and there is only one `proposals.yaml` output file. So, for the test `hm4` `A sparse new proposal` to pass, the entry for `20_adensenewproposal` must be deleted from `proposals.yaml`, and for the test test `hm5` `A dense new proposal` to pass, the entry for `20_asparsenewproposal` must be deleted from `proposals.yaml`. The tests are passing for me when I delete the respective outputs – would you prefer I only provide one test in `test_helpers.py` and hence just one added proposal in proposals.yaml` ?

sbillinge commented:
haha, the tests have to pass with no human present!   See if you can figure
out a way.  The fact that the adder helpers update the dbs actually make
this testing structure brittle, so we may want to think of a more robust
way.  I am open to suggestions, but it has to test effectively and not
reduce the efficacy of the tests.

On Sat, Jun 6, 2020 at 3:13 PM Adiba Ejaz <notifications@github.com> wrote:

> @sbillinge <https://github.com/sbillinge> If I'm not wrong, it's failing
> the tests because two tests for two different proposals have been
> specified, and there is only one proposals.yaml output file. So, for the
> test hm4 A sparse new proposal to pass, the entry for 20_adensenewproposal
> must be deleted from proposals.yaml, and for the test test hm5 A dense
> new proposal to pass, the entry for 20_asparsenewproposal must be deleted
> from proposals.yaml. The tests are passing for me when I delete the
> respective outputs – would you prefer I only provide one test in
> test_helpers.py and hence just one added proposal in proposals.yaml` ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/regro/regolith/pull/372#issuecomment-640105577>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAOWUOCFG3WRPTGFNAW2KDRVKIMHANCNFSM4NOLRPHA>
> .
>


-- 
Professor Simon Billinge
Columbia University

adibaejaz commented:
@sbillinge I have updated this to one test and created a new issue #415 to deal with this aspect of the testing!

sbillinge commented on news/a_proposal.rst:
```diff
@@ -0,0 +1,14 @@
+**Added:**
+* adder helper to add a new proposal to the proposals collection. proposals
```
we just need the first sentence here.

adibaejaz replied:
done

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -0,0 +1,190 @@
+                       help="Anything to note", default = ''
+                       )
+
+    #subpi.add_argument("-t", "--target_collection",
```
please delete

adibaejaz replied:
done!

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -0,0 +1,190 @@
+    subpi.add_argument("--cppflag", help="Current and pending form (True or False). Defaults to True)",
+                       default = True
+                       )
+    subpi.add_argument("--other_agencies", help="Other agencies to which the proposal has been "
```
I think this needs to be a string, so default should be ``"none"`` not ``False``

adibaejaz replied:
The schema says it can be either a string or a boolean – would you prefer I edit this in the schema as well to only string?

sbillinge replied:
I think I wouldn't change the schema in this case.  If you give a default of ``"None"`` (the string) I think that is what we want to print into the cpp form though which seems easier (I think it should be capitalized.....sorry!)

adibaejaz replied:
done!

sbillinge replied:
sorry, I may not have been so clear.  What I meant was that if none of the programs in regolith (e.g., builders etc.) uses "duration" we could remove it from the schema.  To check this using pycharm highlight ``duration`` and do CTL+SHIFT+F and it will show you everywhere in the project that the word is found (assuming you have opened regolith as a project).  This is quick if you have your pycharm set up right.

Then in the adder, I was thinking we could offer the user to add a begin date and a duration, or to add a begin date and an end date, but not both (or to avoid frustration, allow them to add both if they are consistent).  Make sure the help messages make this clear.

Does it make sense?

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -0,0 +1,190 @@
+                       )
+    subpi.add_argument("--begin_date", help="The begin date for the proposed grant "
+                                          "in format YYYY-MM-DD.",
+                       default = 'tbd'
```
I think if you are handling this below, as you seem to be, it can be left with no default, which will return ``None`` and then you handle it with ``if not rc.begin_date:`` which makes the intent a bit clearer in the code.

adibaejaz replied:
That's true - I added defaults everywhere for consistency but they make the code less readable. Will remove them for the `begin_date`, `due_date`, and `duration`

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -0,0 +1,190 @@
+                                          "in format YYYY-MM-DD.",
+                       default = 'tbd'
+                       )
+    subpi.add_argument("-d", "--duration", help="Duration of proposal in months",
```
same as above.    Or, if we use a default it is because we want to pass the value directly into the right field in the document, in which case it has to be the right type.

adibaejaz replied:
done

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -0,0 +1,190 @@
+    subpi.add_argument("-a", "--authors", nargs="+",
+                       help="Other investigator names", default = []
+                       )
+    subpi.add_argument("-c", "--currency",
```
this is good use of defaults here.....nice

adibaejaz replied:
thank you!

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -0,0 +1,190 @@
+                       default = 'USD'
+                       )
+    subpi.add_argument("-p", "--pi",
+                       help="ID of principal investigator. Defaults to"
```
you will need a whitespace at the end of this line for this to display correctly

adibaejaz replied:
done!

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -0,0 +1,190 @@
+    subpi.add_argument("--other_agencies", help="Other agencies to which the proposal has been "
+                                                "submitte. Defaults to False", default = False
+                       )
+    subpi.add_argument("-i", "--institution", nargs="+",
```
thinking about it, I think this has to be a string, not a list.

adibaejaz replied:
You mentioned in an earlier comment that a tweak to the schema in an emerged PR defined institutions as a list - is this no longer the case?

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -0,0 +1,190 @@
+                       default = 0
+                       )
+    subpi.add_argument("--months_summer", help="Number of working months in the summer to be"
+                                                "stated on the current and pending form"
```
I think you will have to work on whitespace in all your help messages.  Run the code locally if you can (I can send a test proposal file if you need) to check for these things (pytest won't show this up)

adibaejaz replied:
My bad! Found this error in multiple places by running `regolith helper a_proposal` again and reading the help carefully

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -0,0 +1,190 @@
+    subpi.add_argument("-f", "--funder", help="Agency where the proposal is being submitted",
+                       default = ''
+                       )
+    subpi.add_argument("-n", "--notes",
```
this should allow for lists and return an empty list

adibaejaz replied:
done!

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -0,0 +1,190 @@
+        if rc.due_date != 'tbd':
+            pdoc.update({'due_date': date_parser.parse(rc.due_date).date()})
+        else:
+            pdoc.update({'due_date': rc.due_date})
```
we need to make sure the schema allows string or datetime (I think it does, but make sure it works)

adibaejaz replied:
The schema as of now specifies string. Should I update it? Further, how do we input a `datetime` object in the CLI?

sbillinge replied:
Thanks Adiba, yes, please add ``date`` (rather than datetime here I think) and make it ``anyof_type`` to the schema (unrelated, but please can you look at this in the meetings schema also, I think it needs fixing there too).  The CLI always takes a string so we have to convert it in the code.  Please look at one of the other converters to see how to do it.

adibaejaz replied:
`meetings` doesn't seem to have a schema in schemas.py, but there is a `due_date` input in the schema for `proposalReviews` which currently only allows string inputs, should I fix this to `"anyof_type": ["string", "date"]` as well?
Also, on the conversion, I am a little unsure what additional conversion you'd like me to add. I already have the functionality to convert a string of type `"YYYY-MM-DD"` to a `datetime` object. Should I also add the functionality to convert a string `"(YYYY, MM, DD)"` or `"datetime.date(YYYY, MM, DD)"` to a datetime object?

sbillinge replied:
yes please!  Its always good to clean code when you find/touch it, as long as it doesn't break anything.  

We will need a schema for meetings now it is fairly stable, but that would be for a different issue and a different PR.  You could help by posting that as a new issue on regolith and someone can volunteer for it in the future.  

For the date formatting, No. for the CLI we will enforce dates in the YYYY-MM-DD form.  If there is a gui somewhere in the future it can handle all the different ways people do it.

adibaejaz replied:
done!

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -0,0 +1,190 @@
+            else:
+                pdoc.update({'end_date': 'tbd'})
+        else:
+            pdoc.update({'duration': rc.duration,
```
is duration in the schema or just end_date?  It seems (as we have discussed before) that there is redundancy here.  Maybe we should think about this.

adibaejaz replied:
`duration` is in the schema explicitly – but I was wondering this too. I think it might be helpful to maintain it for list helpers (for e.g. all proposals of `duration` `x`) to avoid the extra computation, or in a situation where we know the `duration` but not the `begin_date`

adibaejaz replied:
not sure how likely it is that such a situation would arise though

adibaejaz replied:
fixed!

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -0,0 +1,190 @@
+            pdoc.update({'duration': rc.duration,
+                         'end_date': 'tbd'
+                         })
+        if rc.funder:
```
I think this has a default set so you don't need an if - else clause, right?  Same for all the below.  Just load ``rc.funder`` and it will ether be what the user typed or the default if the user didn't type.

adibaejaz replied:
done

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -0,0 +1,190 @@
+        rc.client.insert_one(rc.database, rc.coll, pdoc)
+
+        print(f"{key} has been added in {TARGET_COLL}")
+        #print(f"{key} has been added in {rc.target_collection}")
```
pls clean

adibaejaz replied:
done

sbillinge commented on tests/test_helpers.py:
```diff
@@ -17,6 +17,10 @@
       "--date", "2020-04-29", "-c", "afriend", "-d", "more work",
       "-m", "ascopatz", "-g", "SymPy-1.1"],
      "20ly_newprojectum has been added in projecta\n"),
+    (["helper", "a_proposal", "a new proposal","100.0","To destroy numbers",
```
maybe load two things for things that take lists for slightly better coverage?

adibaejaz replied:
done!

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -144,33 +134,27 @@ def db_updater(self):
                 pdoc.update({'end_date': 'tbd'})
         else:
-            pdoc.update({'duration': rc.duration,
+            pdoc.update({'duration': 'tbd',
```
Please could you search for anywhere in the code that uses duration?  if not we may remove it from the db?  Do you think it would be less confusing in the adder to have ``begin_date`` + ``duration`` or ``begin_date`` + ``end_date``, or both maybe, with logic in the code that throws an exception if both are specified and inconsistent?

adibaejaz replied:
done - I have removed `duration`, so now the code takes in `begin_date` and `end_date` and raises a `ValueError` exception if both are specified and inconsistent

adibaejaz replied:
can't comment above but capitalisation's changed!

sbillinge commented:
@adibaejaz  this seems to be failing CI, could you take a look?

adibaejaz commented:
@sbillinge I have updated some fields to `required: False` and added`begin_date` and `end_date` to `schemas.py`. I have also changed the dates behaviour to allow the user to enter a `begin_date` and either an `end_date` or `duration`, or all 3 provided they are consistent (any inconsistencies lead to a `ValueError`).

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -128,6 +131,12 @@ def db_updater(self):
                 if begin_date > end_date:
                     raise ValueError("begin_date can not be after end_date")
+                if rc.duration:
+                    expected_duration = 12*relativedelta(end_date, begin_date).years + relativedelta(end_date, begin_date).months
```
how will this behave if there is not an integral number of months?

adibaejaz replied:
`relativedelta(date1, date2).months` returns the floor of the difference in months, whereas `begin_date + relativedelta(months=rc.duration)`only supports integral values for `month` as defined in the documentation. so in this equality, I should probably add a floor(float(rc.duration))... unless we want to be comparing exact differences? i.e. the difference between `begin_date` and `end_date` should also be a float?

sbillinge replied:
not sure I have a preference as long as it works.  Maybe think about some edge cases, how you would want them to behave, and then how they do behave given the different options.  This is where TDD is helpful because we test for the edge cases, then make the code work.  One way to do this is if we break this out into a function that we put into dates.py.  It may make sense or it may not.....

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -128,6 +131,12 @@ def db_updater(self):
+                if rc.duration:
+                    expected_duration = 12*relativedelta(end_date, begin_date).years + relativedelta(end_date, begin_date).months
+                    if expected_duration != int(rc.duration):
+                        raise ValueError(f"Entered duration {rc.duration} is inconsistent with"
```
I think this is tmi.   Just write  "ERROR: please rerun specifying a duration OR an end-date but not both"

adibaejaz replied:
done

sbillinge commented on regolith/helpers/a_proposalhelper.py:
```diff
@@ -157,9 +169,9 @@ def db_updater(self):
         else:
             pdoc.update({'pi': rc.pi_id})
-        pdoc.update({'status': 'inprep'})
+        pdoc.update({'status': 'in-prep'})
```
I would rather have inprep.  If this means a schema change, let's do the shema change.  It is always a headache to keep track of uppercase letters and dashes and underscores in allowed values, so my go-to rule is not use them unless really necessary.

adibaejaz replied:
Ok! I will add `inprep` to `ALLOWED_STATI` but leave `in-prep` in for backwards compatibility

sbillinge replied:
Thanks!  Please remove ``in-prep``  Your decision was a good one but in-prep is mine so there are no backwards compatibility issues further back than when I took over....

adibaejaz replied:
done!

sbillinge commented on tests/outputs/a_proposal/proposals.yaml:
```diff
@@ -19,10 +19,10 @@
     - this is a sample added proposal
   pi: scopatz
-  status: inprep
+  status: in-prep
```
see above

adibaejaz replied:
done

sbillinge replied:
inprep

adibaejaz commented:
@sbillinge I have added some more functionality that lets us compare `duration` and `end_date - begin_date` up to the exact day (to better handle fractional durations). There is another function below it in a comment block that could be used, although it's not as precise, and comments explaining why I use certain expressions that I will clean up in the final version! I have tried it with several edge cases and thought I'd push it as a rough version of something we could possibly include in `dates.py`.

sbillinge commented:
thanks @adibaejaz great job!  pls could you put in a PR to finish the prum for this?  add an end_date and set all milestones and the prum status to finished

adibaejaz commented:
@sbillinge  yes I will! thank you Professor this is very exciting

sbillinge commented:
I will test it out.  If I find anything as I use it I will post issues.  This usually happens but it is a good sign....it means it is getting used!

sbillinge commented:
look how pretty this is..... :)
[image: image.png]


On Wed, Jun 17, 2020 at 8:19 PM Adiba Ejaz <notifications@github.com> wrote:

> @sbillinge <https://github.com/sbillinge> yes I will! thank you Professor
> this is very exciting
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/regro/regolith/pull/372#issuecomment-645693644>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAOWUMDE422FDFJDHPTFFTRXFMPNANCNFSM4NOLRPHA>
> .
>


-- 
Professor Simon Billinge
Columbia University
